File: Data Structures/Python/Hashing.py
#------------------------------------------------------------
max_size = 10

#------------------------------------------------------------
class Hash:

      def __init__(self):
            self.__hashset = [-1]*max_size
      
      #------------------------------------------------------------
      def insert(self,data):
            index = data % 10
            index = self.find_index(index)
            self.__hashset[index] = data

      #------------------------------------------------------------
      def display(self):
            for i in range(max_size):
                  print(i," :: ",self.__hashset[i])

      #------------------------------------------------------------
      def find_index(self,index):
            if self.__hashset[index] == -1:
                  return index
            else:
                  if index < max_size-1:
                        return self.find_index(index + 1)
                  else:
                        index = 0
                        return self.find_index(0)
      
      #------------------------------------------------------------
      def search(self,element):
            index = element % 10
            flag = 0
            while self.__hashset[index] != -1:
                  if self.__hashset[index] == element:
                        flag = 1
                        break
                  else:
                        if index < max_size-1:
                              index += 1
                        else:
                              index = 0
            
            if flag == 1:
                  print('Element Found..!')
            else:
                  print('No such element..')

#------------------------------------------------------------
if __name__ == '__main__':

      myhash = Hash()
      n = int(input('Enter Number Of Elements :: '))
      #------------------------------------------------------------
      for i in range(n):
            data = int(input('Enter Data :: '))
            myhash.insert(data)
      
      #------------------------------------------------------------
      myhash.display()

      #------------------------------------------------------------
      element = int(input('Enter the element to be found :: '))
      myhash.search(element)

File: Data Structures/C++/Hashing.cpp
#include<iostream>

#define SIZE 10

//----------------------------------------------------------
class Hash{
	private : 
		int hashset[SIZE];
	public :
		Hash(){
			for(int i=0 ; i<SIZE ; i++){
				this->hashset[i] = -1;
			}
		}
		void map(int data);
		void display();
		int find_location(int index);
		void search(int data);
};

//----------------------------------------------------------
//both recursive and iterative approach
void Hash::map(int data){
	int index = data % 10;
	index = find_location(index);
	this->hashset[index] = data;
	//Iterative
	/*while(1){
		if(this->hashset[index] == -1){
			this->hashset[index] = data;
			break;
		}
		else{
			index++;
			if(index >= SIZE){
				index = 0;
			}
		    }	
	}
	*/
	
}

//----------------------------------------------------------

int Hash::find_location(int index){
	if(this->hashset[index] == -1){
		return index;
	}
	else{
		if(index < SIZE-1){
			return find_location(index + 1);
		}
		else{
			return find_location(0);
		}	
	}
}
//----------------------------------------------------------

void Hash::search(int data){
	int flag = 0;
	int index = data % 10;
	while(this->hashset[index] != -1){
		if(this->hashset[index] == data){
			flag = 1;
			break;
		}
		else{
			if(index < SIZE-1){
				index ++;
			}
			else{
				index = 0;
			}
		}
	}
	
	if(flag == 1){
		std::cout<<"\nElement Found..!\n";
	}
	else{
		std::cout<<"\nElement Not In Hash Table\n";
	}
}

//----------------------------------------------------------
void Hash::display(){
	for(int i=0 ; i<SIZE ; i++){
		std::cout<<i<<" :: "<<this->hashset[i]<<"\n";
	}
}

//----------------------------------------------------------
int main(){
	
	int n;
	int data;
	Hash myhash = Hash();
	std::cout<<"Enter number of elements :: ";
	std::cin>>n;
	
	//----------------------------------------------------------
	for(int i = 0 ; i<n ; i++){
		std::cout<<"Enter Data :: ";
		std::cin>>data;
		myhash.map(data);
	}
	
	myhash.display();
	
	//----------------------------------------------------------
	std::cout<<"Enter Data to be Found :: ";
	std::cin>>data;
	myhash.search(data);
	return 0;
}
